protocol/l1-upgrades.md

```diff
@@ -0,0 +1,305 @@
+`Populator` contract being temporarily set as the implementation in order to set or modify any
+values as required. This has two significant benefits. First it removes the need to keep a one-time
+`initialize()` function in the run-time logic. Secondly it enables a clean separation of concerns
+between contracts which are newly deployed (requiring all values to be set), and contracts being
+upgraded (requiring only new values to be set).
+
+# Problem Statement + Context
+
+Our upgrades have the following issues:
+
+1. They are divorced from the implementation, meaning the work of writing the upgrade script occurs
+   in `superchain-ops` apart from the work of writing the updated code in the `optimism` monorepo.
+2. The act of writing the upgrade script is onerous, requiring:
+   1. A bespoke script with custom assertions.
+   2. A bespoke Validations.md document
+3. The work required to perform the upgrade is difficult for signers, as they must work through the
+   Validations.md document. This work has taken up to an hour in the past.
+4. The effort described above will scale linearly with the number of OP Chains.
+
+# Proposed Solution
+
+Each new release will have its own OP Contracts Manager which can perform the following distinct
+actions:
+
+1. Deploy new proxies for a new OP Chain (this work is already covered by [link to Blaine's doc]())
+1. Upgrade all existing OP Chains.
+1. Deploy any newly added superchain-shared contracts.
+1. Upgrade existing superchain-shared contracts.
+
+Note that the above functionality also makes for a good set of milestones, each of which can and
+should be implemented incrementally.
+
+## Declarative storage management
+
+Taking inspiration from Chugsplash, we wish to separate the concern of what bytecode to run, and
+what to write to storage. However we also wish to avoid using the current `StorageSetter` pattern,
+which simply provides logic to "write anything anywhere" in storage.
+
+In order to achieve this, each contract will be refactored to inherit from a `__Layout` contract. By
+way of example, we take an initializable version of `SimpleStorage`, and refactor it into the set of
+contracts which would be required to specify a hypothetical Magma release, which adds a new variable
+to the contract.
+
+```solidity
+contract SimpleStorage is Initializable {
+  uint public foo;
+  bytes32 public bar; // new variable added for the Magma release.
+
+
+  /// @notice initialize is called whether for a fresh system or when upgrading an existing system.
+  function initialize(uint _foo, bytes32 bar) public initializer {
+    // chains being upgraded need to read the current value of foo, then write it again.
+    foo = _foo;
+    bar = _bar;
+  }
+}
+```
+
+This `SimpleStorage` will become:
+
+```solidity
+abstract contract SimpleStorageLayout {
+  uint public foo;
+  bytes32 public bar;
+}
+
+contract SimpleStoragePopulator is SimpleStorageLayout {
```

Comment:
What would a more complex example look like, where `SimpleStorage` inherits from 2 or more base contracts that have their own storage layouts, epecially when they are deps like OZ where we might not have control over the storage layout? I suspect this gets a bit messer and perhaps not doable.

The first alternative that comes to mind is to remove `SimpleStorageLayout` and do `SimpleStoragePopulator is SimpleStorage` and add the populator methods. The populator has a lot of extra code but that is mainly harmless

Reply:
> What would a more complex example look like, where SimpleStorage inherits from 2 or more base contracts that have their own storage layouts, especially when they are deps like OZ where we might not have control over the storage layout? I suspect this gets a bit messer and perhaps not doable.

I expect they would look a lot like the state var declaration section of our existing contracts. Unless I've missed something, the only external deps our L1 contracts currently inherit from are `Initializable` and `OwnableUpgradable`. This proposals removes the former, and we should also move away from using the Ownable pattern. Gaps and spacers can be used as needed. 

> The first alternative that comes to mind is to remove `SimpleStorageLayout` and do `SimpleStoragePopulator` is `SimpleStorage` and add the populator methods.

The benefit of the proposed approach is that we don't get the additional buffer below the contract size limit, that we get form moving the `initializer` into a contract with no other logic. However that is a 'nice to have' beyond the goals, so I like this as a simplification.

Reply:
Outdated